packages/fxa-auth-server/lib/payments/stripe.js

```diff
@@ -856,6 +856,106 @@ class StripeHelper {
       this.webhookSecret
     );
   }
+
+  /**
+   * Accept a string ID or resource object, return a resource object after
+   * retrieving (if necessary)
+   *
+   * @template T
+   * @param {string | T} resource
+   * @param {string} resourceType
+   *
+   * @returns {Promise<T>}
+   */
+  async expandResource(resource, resourceType) {
+    return typeof resource === 'string'
+      ? this.stripe[resourceType].retrieve(resource)
+      : resource;
+  }
+
+  /**
+   * Extract invoice details for billing emails
+   *
+   * @param {Invoice} invoice
+   */
+  async extractInvoiceDetailsForEmail(invoice) {
+    let uid = '',
+      email = '',
+      invoiceNumber = '',
+      invoiceDate,
+      invoiceTotal = 0,
+      cardType = '',
+      lastFour = '',
+      nextInvoiceDate,
+      productId = '',
+      planId = '',
+      planName = '',
+      planEmailIconURL = '',
+      planDownloadURL = '';
+
+    try {
+      if (invoice && typeof invoice === 'object') {
+        // Dig up & expand objects in the invoice that usually come as just IDs
+        const { plan } = invoice.lines.data[0];
+        const [product, customer, charge] = await Promise.all([
+          this.expandResource(plan.product, 'products'),
+          this.expandResource(invoice.customer, 'customers'),
```

Comment:
Would this be in the customer cache?

Reply:
No - our customer cache is keyed on FxA uid & email. We only have a customer ID here.

Reply:
We might want to add a follow-up issue to start caching customers on Stripe ID, which would let us use cache here. But, I also kind of think these emails would be a decent spot to have fresh data?

Reply:
The Redis cache should be up to date, if its not then we have a bug.